### PR TITLE
Add support for Python 3.12

### PR DESCRIPTION
**Pull request by lbianchi-lbl**
## Summary/Motivation:

Python 3.12 is now the most recent stable version and we should ensure it's supported and tested.

## Changes proposed in this PR:

- Add Python 3.12 to CI job matrices and docs
- Update default Python version in CI to 3.11

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by jsiirola — APPROVED** [review]

**Review by blnicho — APPROVED** [review]

File: idaes/core/dmf/commands.py [hunk 1]
```diff
@@ -391,7 +391,9 @@ def cat_resources(path, objects=(), color=True):
     # get all resources,
     # display any that match an object as a prefix
     for r in d.find():
-        for oid in unmatched:
+        # iterate on a copy of the set to avoid
+        # mutating an object being iterated on
+        for oid in set(unmatched):
```
Comment (jsiirola) [inline idaes/core/dmf/commands.py:396]: I would make this a `list` and not a `set`: list is faster, doesn't require hashability, and most importantly, is deterministic.  I think I would have gone with:
```python
unmached = list(objects)
for r in d.find():
    for i, oid in enumerate(unmached):
        if not oid:
            continue
        if r.uuid.startswith(oid):
            unmached[i] = None
            # ...
```

**Review by jsiirola — APPROVED** [review]
This is fine, although there is a potential nondeterminism in the `dmf/commands.py` that may (or may not) be an issue.

**Comment by lbianchi-lbl** [comment]
Definitely orthogonal to the main purpose of this PR, but after updating Pylint (which was needed for 3.12 compatibility), I figured out a way to format its log so that it's nicely rendered by GitHub:

image

File: idaes/core/dmf/commands.py (hunk 1 above)
Comment (lbianchi-lbl) [reply idaes/core/dmf/commands.py]: As a note for posterity (?): I've moved the Pylint-related changes (only incidentally due to Python 3.12), including the one this comment refers to, to their own PR (#1351, now merged as 57cf0a7f31f3b5aafabfe103c5af8d028bb69b37)

**Pull request merged**